[알린]Store step8 PR

jinios opened the pull request:
마지막 단계 구현하였습니다!!
네트워크 연결성을 확인하는 부분은 상태바 색상변경과 깜빡이는 애니메이션 적용으로 응용했습니다 : )

jinios committed:
Import framework - Alamofire

jinios committed:
Feat: Add NetworkManager class #21

jinios committed:
Add icon images

jinios committed:
Refactor: Modify error log, add annotations

jinios committed:
Feat: Check network reachability #21

- Check status of network
- Change the color of status bar as network status changes
- Add blink animation on status bar

jinios committed:
Refactor: Make HashDataSetter file #21
- Split HashDataSetter class from DetailHash.swift

jinios committed:
Feat: Add UnreachableViewController #22

- To Show a view that indicates 'Currently there's no network
connection'

jinios committed:
Add UnreachableViewController in storyboard #22

jinios committed:
Feat: Add Local description - Unreachable Network #22

jinios committed:
Feat: Show different color on status bar  #23,#21

- Show different color on status bar by network reachability
- Add reachable variable in NetworkManager
- Modify blink animation by reachability status

jinios committed:
Refactor: Delete URLDataProtocol

jinios committed:
Feat:Check Reachability before dataTask starts #21

- Execute tryDownload() before dataTask begins to check network status
- Load data from local or from cache if the network is unreachable
- Notification named .sectionComplete only sends Category as a userInfo
- (Tableview only reloadsections when it receives newly loaded data)

jinios committed:
Modify function - download to tryDownload #21

jinios committed:
Feat: Modify reloading tableview #21, #22

- Apply status of NetworkManager to a viewController scenario
- Add toUnreachableView()
- If the status is unreachable, the error view will be pushed and shown to a user when a cell tapped
- Separate the scenario of next view controller by network reachability
- Modify the way of reloading on the  tableView: Changed to reloadSections

jinios committed:
Refactor: Modify parameter of reloadSections()

jinios committed:
Refactor: Add custom colors to asset

- Add custom colors to asset catalog

jinios committed:
Feat: Change status bar style by reachability #23

jinios committed:
Docs: Update docs - step8

godrm commented on StoreApp/StoreApp/AppDelegate.swift:
```diff
@@ -41,6 +72,29 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+
+    var statusBarView: UIView? {
+        return value(forKey: "statusBar") as? UIView
+    }
```
이 코드가 어떤 원리로 동작하는 걸까요? 설명을 좀 해주세요 ^^

godrm commented on StoreApp/StoreApp/Model/HashDataSetter.swift:
```diff
@@ -0,0 +1,36 @@
+                    let parsedData = try JSONDecoder().decode(ItemHashData.self, from: data)
+                    handler(parsedData)
+                } catch {
+                    print("Parse Error!!!")
```
tryDownload() 나 set() 에서 Handler가 있는데 비정상적인 경우에는 print()만 하고 있네요.
상위 모듈로 handler를 통해서 전달해서 다음 상황에 대비할 수 있도록 처리해보세요.

godrm commented on StoreApp/StoreApp/Model/NetworkManager.swift:
```diff
@@ -0,0 +1,45 @@
+                self.reachable = true
+            }
+            NotificationCenter.default.post(name: .reachabilityChanged, object: self, userInfo: ["reachabilityStatus":status])
+            print("이전 연결: \(self.reachabilityManager?.previousFlags.rawValue))")
```
개발과정에서 로그를 확인하는 건 좋은데 저장소에 올릴때는 print()를 남겨놓는게 좋은 습관은 아닙니다. 
브레이크 포인트를 적극 활용해보세요.

jinios replied:
저번에 말씀해주신게 있어서 저도 최대한 지양하고 브레이크 포인트를 활용하고있긴한데, 이 부분이 비동기로 돌아가기도하고, 시뮬레이터가 아닌 실제 디바이스에서 돌아갈때 디버거에 찍히는걸 바로 확인하려고 이렇게 했습니다. 나중에 머지되고 최종확인하고나서 다 정리하려고했지만, 최대한 print()는 안쓰도록하겠습니다!! ㅠㅠ

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -19,61 +20,73 @@ class StoreViewController: UIViewController {
         #if DEBUG
-        self.deleteCache()
+        // self.deleteCache()
         #endif
```
이 코드는 어떤 용도였을까요? 
`#if DEBUG`는 어떤 경우에 사용하는지 정리해보세요.

jinios replied:
`#if DEBUG` ~ `#endif` 사이에 코드를 넣으면 해당 코드는  Build configuration 이 Debug모드일때만 해당 코드가 실행됩니다. 

이전 스텝에서 캐시 디렉토리에 이미지를 저장했을때 이미지가 저장되는지/저장이 되었다면 캐시에 저장된 데이터를 사용하는지를 확인하고자 넣은 코드입니다.

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -19,61 +20,73 @@ class StoreViewController: UIViewController {
+            }
+        } else {
+            print("RootView - dataReload. Not reachable")
+        }
```
dataReload 노티를 받았을 때 ViewController가 판단해서 데이터를 다른걸 채우는건가요?
ViewController는 네트워크 상태를 알 필요없이, StoreItems에 (혹은 네트워크 관리하는 객체가 있다면...) 요청하면 이 동작을 StoreItems 에서 처리하는건 어떤가요?

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -19,61 +20,73 @@ class StoreViewController: UIViewController {
                 }
                 let files = try ImageSetter.fileManager.contentsOfDirectory(atPath: cacheURL.path)
-                print("all files in cache after deleting images: \(files)")
             }
```
deleteCache()는 ViewController에 있어야 하는 메소드가 맞나요?

jinios replied:
deleteCache()가 위의 `#if DEBUG`블럭 내에서 실행되는 코드라서, 같은 파일 내에 있는게 낫다고 생각했었습니다ㅠㅠ... 
수정하면서 필요없다면 지우고 필요하다면 ImageSetter로 옮기겠습니다!

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -93,14 +106,17 @@ extension StoreViewController: UITableViewDataSource {
-            self.navigationController?.pushViewController(nextVC, animated: true)
+        } else {
+            self.toUnreachableView()
         }
```
여기도 DataSource가 네트워크 상태를 모르게 주어진 데이터로만 처리하도록 개선해보세요.

jinios replied:
JK질문있습니다!! 네트워크 연결성을 확인하는 부분이 StoreItems나 DetailHash와 같은 모델로 제한하는 흐름으로 통일해야하는건가요?
(DetailHash는 상품의 hash값으로 데이터를 받아와서 저장하는 모델입니다)

그러면 해당 모델에서 네트워크를 체크한 결과에따라 뷰컨트롤러가 동작할수있도록 알려주는 방식도 안되는건가요?_? (notification을 보낸다든지...하는 방식으로요)

godrm replied:
ViewController가 네트워크 상태에 따라 모델을 다르게 접근하기 보다는
네트워크 상태를 관리하는 계층이 어디에서 동작할지 고민해보는 게 좋습니다.
ViewController는 상태를 모르고, 모델에서 값이 올때마다 뷰에 반영하도록 단순하게 작성하는 게 좋습니다.
VC --> Network Manager --> 상태확인후 Request 또는 저장된 데이터 처리 --> 모델 변경 --> 노티 --> VC 화면 변경 흐름을 상상해봤습니다.

godrm reviewed (commented):
다음 피드백을 확인해보세요! 
마지막이니까 더 힘내서 으싸!!

jinios committed:
Refactor: Modify flow of checking reachability

- Modify flow: DataSetter classes check network reachability
- Delete rechability checking codes from StoreViewController
- Add logic in error handling block in HashDataSetter
- DetailHash class: Add features to check whether ItemHashData is nil or not- DetailHash class: Rename variable

jinios commented:
피드백 주신 사항 수정했습니다! 

마지막에 알려주신 흐름과 똑같진 않지만, 현재 StoreItems모델에 값을 넣어줄때 네트워크 연결성을 확인하고 URLSession.dataTask를 실행하는 부분을 이용했습니다.
- 정보를 로딩하는객체(Setter 모델 객체들) 내부에서만 네트워크 연결 여부를 판단하고, 그 여부에 따라 핸들러에 넘겨주는 값을 다르게 설정했습니다. 
- 따라서 뷰컨트롤러에서는 모델의 값이 nil인지 판단하여 어떤 화면을 표시할지 결정하는 방식으로 수정해보았습니다.

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -20,9 +20,6 @@ class StoreViewController: UIViewController {
-        #if DEBUG
-        // self.deleteCache()
-        #endif
         NotificationCenter.default.addObserver(self, selector: #selector(dataReload), name: .reachabilityChanged, object: nil)
```
ViewController 가 .reachabilityChanged 상태를 옵저빙 해야 하는걸까요? 
상태 변화를 감시할 게 아니라 그냥 데이터 모델이 바뀌었는지만 감시하면 되는거 아닐까요?

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -106,19 +82,19 @@ extension StoreViewController: UITableViewDataSource {
+                    self.navigationController?.pushViewController(nextVC, animated: true)
+                } else {
+                    self.toUnreachableView()
                 }
```
여기 있는 if let 옵셔널 바인딩은 guard let 으로 바꾸는게 더 깔끔해보일 것 같습니다.

godrm commented on StoreApp/StoreApp/StoreViewController.swift:
```diff
@@ -60,23 +53,6 @@ class StoreViewController: UIViewController {
-
     private func toUnreachableView() {
         if let unreachableVC = self.storyboard?.instantiateViewController(withIdentifier: "unreachableViewController") as? UnreachableViewController {
             self.navigationController?.pushViewController(unreachableVC, animated: true)
```
화면을 못 봐서 그런데 UnreachableViewController 는 어떤 뷰 컨트롤러 화면인가요?

godrm commented on StoreApp/StoreApp/Model/HashDataSetter.swift:
```diff
@@ -10,25 +10,25 @@ import Foundation
                 }
             } else {
-                print("데이터 에러!")
+                handler(nil)
```
`handler(nil)` 을 호출하면 받는 쪽에서 바뀐게 있나요?
어디서 하는지 못 찾겠습니다. 어떤 동작을 해야할까요?

jinios replied:
핸들러에 nil을 넘기면 DetailHash의 itemHashData 프로퍼티가 nil로 세팅됩니다. 
테이블뷰 셀은 해당 값을 프로퍼티로 가지고만 있고, 상품상세페이지를 담당하는 뷰컨트롤러가 해당 
 셀에 있는 itemHashData값을 가지고 뷰를 구성하게되어있어서, 상위모듈에서 nil을 체크해서 상품 상세페이지 뷰를 보여줄지 말지 결정할 수 있습니다.

jinios committed:
Feat: Add SystemOrange color to asset

jinios committed:
Refactor: Delete print log

jinios committed:
Refactor: Modify flow to set StoreItems

- StoreItems class has the observer to check reachability, this class
sets its data as reachability changes
- To reload tableView, only model class tracks reachability with
notification sent from NetworkManager

- StoreItems: Add observer to track reachability
- StoreViewController: Delete Notification observer which tracks
reachability

jinios committed:
Refactor: Delete print log

jinios committed:
Fix: Modify logic on stackview

- Fix logic implement stackview of badge and prices
- Add prepareForReuse()

jinios commented:
피드백 주신 부분수정했습니다! ㅠㅠㅠㅠ
- 연결성이 바뀌는걸 체크하는 notification옵저버를 StoreItems객체로 옮기고 뷰컨트롤러는 StoreItems객체의 세팅유무만 판단해서 테이블뷰를 로딩하는 흐름으로 변경했습니다.

godrm commented:
View 에서 ViewController > Network Manaer > Model 까지 흐름과
Model 에서 변경된 사항을 ViewController > View 로 업데이트 하는 흐름을 항상 구분하도록 설계하세요.
그래야 데이터나 네트워크에 변화가 있어도 화면을 업데이트 하는 흐름은 덜 영향을 받습니다. 
수고하셨습니다.
